### ISSUE TITLE
Extract Variable refactoring screws up code

### ISSUE THREAD

**Issue by automatensalat** [nicoespeon/abracadabra#88]
## Describe the bug

When I try to apply the Extract Variable refactoring, the generated code is messed up, resulting in not compilable code. See below. 

## How to reproduce

Consider this code: 
```ts
function test(arg: string[]) {
	console.log(arg.length);
	const a = arg.length + 1;
	const b = arg.length % 2;
	return { a, b };
}
```
Put the cursor on the `length` part of `arg.length` (e.g. on the "n"). Execute Extract Variable refactoring.

The refactoring generates this code: 
```ts
function test(arg: string[]) {
	conconst { length } = arg;
    sole.log(arglength
	const a = arglength1;
	const b = arglength2;
	return { a, b };
}
```

## Expected behavior

The expression is extracted to a single variable and each occurence is replaced with the created variable. No syntax error is created. 
E.g. something like this:
```ts
function test(arg: string[]) {
        const length = arg.length;
	console.log(length);
	const a = length + 1;
	const b = length % 2;
	return { a, b };
}
```

## Additional information

- Version of the extension impacted: 3.2.1

**Comment by nicoespeon** [comment]
Hi @automatensalat 👋 

Thanks for reporting. I'm really curious about this one because I can't reproduce it. I recorded a gif from my editor to show what it looks like with just this code:

![CleanShot 2020-03-14 at 17 59 27](https://user-images.githubusercontent.com/1094774/76691123-c1385280-661d-11ea-93a5-5f39a9c57a28.gif)

So for this specific example, I'm wondering why it doesn't work for you. My main hypothesis is: if it takes time for the editor to trigger the renaming for some reason (it's busy) and you start typing in the meantime, it may mess up the code? Is it what happened?

**Comment by automatensalat** [comment]
Hey @nicoespeon, that's not what happened, the refactoring executes immediately and I don't do any typing in the meantime. I'm going to look into it today to see what happens.

**Comment by automatensalat** [comment]
I believe this has to do with indention. As can be seen in your gif, you use two spaces for indention. The tabs from the example code probably get converted to spaces when saving. The refactoring works when used with spaces or no indention at all; it fails if indented with tabs. 
I'll look into it and try to find a solution.

**Comment by nicoespeon** [comment]
> Edit: The node.loc seems to be screwed up when used with tabs. But that part is Babel, right?

Oooooh, great catch!

Yeah, that's probably Babel since it does the parsing. But I don't get why tabs would screw up the `node.loc`. If it does, there might be another attribute we could use.

You make me realize we never wrote unit tests with tabs inside. I think I can simulate that easily now. Will see if we can find a fix 👍

**Comment by nicoespeon** [comment]
Great news @automatensalat 👉 we have a fix!

Fabien and I did a pair-programming session and we found what the problem was. In short, recast (the library that parses the code on top of Babel) counted 4 characters per tab per default. But VS Code counts one. We fixed that discrepancy and everything works fine now. 

I'll ship that fix by tomorrow 😉

**Comment by automatensalat** [comment]
That's great, thanks! :wave:

**Comment by nicoespeon** [comment]
Fixed in the last release (v3.2.3) 👍
